utility/map_str_enhancer.go

```diff
@@ -4,44 +4,99 @@ import (
 	"github.com/elastic/beats/libbeat/common"
 )
 
-func AddStrWithDefault(m common.MapStr, key string, val *string, defaultVal string) {
-	if val != nil {
-		m[key] = *val
-	} else if defaultVal != "" {
-		m[key] = defaultVal
-	}
-}
-
 func Add(m common.MapStr, key string, val interface{}) {
-
+	if m == nil || key == "" {
+		return
+	}
 	switch val.(type) {
 	case *bool:
 		if newVal := val.(*bool); newVal != nil {
 			m[key] = *newVal
+		} else {
+			delete(m, key)
 		}
 	case *int:
 		if newVal := val.(*int); newVal != nil {
 			m[key] = *newVal
+		} else {
+			delete(m, key)
 		}
 	case *string:
 		if newVal := val.(*string); newVal != nil {
 			m[key] = *newVal
+		} else {
+			delete(m, key)
 		}
 	case common.MapStr:
 		if valMap := val.(common.MapStr); len(valMap) > 0 {
+			for k, v := range valMap {
+				Add(valMap, k, v)
+			}
 			m[key] = valMap
+			if len(valMap) > 0 {
+				m[key] = valMap
+			} else {
+				delete(m, key)
+			}
+		} else {
+			delete(m, key)
+		}
+	case map[string]interface{}:
+		valMap := val.(map[string]interface{})
+		if len(valMap) > 0 {
+			for k, v := range valMap {
+				Add(valMap, k, v)
+			}
+			if len(valMap) > 0 {
+				m[key] = valMap
+			} else {
+				delete(m, key)
+			}
+		} else {
+			delete(m, key)
 		}
 	case []string:
 		if valArr := val.([]string); len(valArr) > 0 {
 			m[key] = valArr
+		} else {
+			delete(m, key)
 		}
 	case []common.MapStr:
 		if valArr := val.([]common.MapStr); len(valArr) > 0 {
 			m[key] = valArr
+		} else {
+			delete(m, key)
 		}
+	case float64:
+		m[key] = common.Float(val.(float64))
 	default:
 		if val != nil {
 			m[key] = val
+		} else {
+			delete(m, key)
+		}
+	}
+}
+
+// MergeAdd modifies `data` *in place*, inserting `values` at the given `key`.
```

Comment:
wrong arg names in comment

Reply:
as a reminder, if you don't replace the maps with `MapStr` in the decoding in this PR (which is ok, this already big), then I think we have to do it as soon as this is merged - is not great to have duplicated functions like this.